File: services/meeting/src/main/java/org/transhub/service/MeetingService.java
package org.transhub.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.transhub.client.FileClient;
import org.transhub.client.UserClient;
import org.transhub.dto.response.UserResponse;
import org.transhub.dto.request.MeetingCreateRequest;
import org.transhub.dto.request.MeetingUpdateRequest;
import org.transhub.dto.request.MemberAddRequest;
import org.transhub.dto.request.MemberUpdateRequest;
import org.transhub.dto.response.ApiResponse;
import org.transhub.dto.response.MeetingMemberResponse;
import org.transhub.dto.response.MeetingResponse;
import org.transhub.entity.Meeting;
import org.transhub.entity.MeetingMember;
import org.transhub.entity.MeetingRole;
import org.transhub.entity.MeetingStatus;
import org.transhub.exception.AppException;
import org.transhub.exception.ErrorCode;
import org.transhub.repository.MeetingMemberRepository;
import org.transhub.repository.MeetingRepository;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class MeetingService {

    private final MeetingRepository meetingRepository;
    private final MeetingMemberRepository meetingMemberRepository;
    private final FileClient fileClient;
    private final UserClient userClient;

    private MeetingResponse mapToMeetingResponse(Meeting meeting) {
        return MeetingResponse.builder()
                .id(meeting.getId())
                .title(meeting.getTitle())
                .description(meeting.getDescription())
                .creatorId(meeting.getCreatorId())
                .audioFileId(meeting.getAudioFileId())
                .status(meeting.getStatus())
                .createdAt(meeting.getCreatedAt())
                .updatedAt(meeting.getUpdatedAt())
                .build();
    }

    private MeetingMemberResponse mapToMemberResponse(MeetingMember member) {
        return MeetingMemberResponse.builder()
                .id(member.getId())
                .meetingId(member.getMeetingId())
                .userId(member.getUserId())
                .role(member.getRole())
                .joinedAt(member.getJoinedAt())
                .build();
    }

    @Transactional
    public MeetingResponse createMeeting(MeetingCreateRequest request, Long creatorId) {
        log.info("Creating meeting with title: {}, creatorId: {}, audioFileId: {}", request.getTitle(), creatorId, request.getAudioFileId());

        // Validate audio file existence using FileClient
        try {
            ApiResponse<Boolean> fileExistsResponse = fileClient.checkFileExists(request.getAudioFileId());
            if (fileExistsResponse == null || fileExistsResponse.getResult() == null || !fileExistsResponse.getResult()) {
                log.warn("Audio file validation failed: file does not exist. ID: {}", request.getAudioFileId());
                throw new AppException(ErrorCode.AUDIO_FILE_NOT_FOUND);
            }
        } catch (AppException ae) {
            throw ae;
        } catch (Exception e) {
            log.error("Failed to connect to file-service for validation", e);
            throw new AppException(ErrorCode.AUDIO_FILE_NOT_FOUND);
        }

        // Check if audio file is already linked to another meeting
        if (meetingRepository.existsByAudioFileId(request.getAudioFileId())) {
            throw new AppException(ErrorCode.AUDIO_FILE_ALREADY_LINKED);
        }

        // Save meeting
        Meeting meeting = Meeting.builder()
                .title(request.getTitle())
                .description(request.getDescription())
                .creatorId(creatorId)
                .audioFileId(request.getAudioFileId())
                .status(MeetingStatus.CREATING)
                .build();
        Meeting savedMeeting = meetingRepository.save(meeting);

        // Add creator as HOST
        MeetingMember member = MeetingMember.builder()
                .meetingId(savedMeeting.getId())
                .userId(creatorId)
                .role(MeetingRole.HOST)
                .build();
        meetingMemberRepository.save(member);

        return mapToMeetingResponse(savedMeeting);
    }

    public MeetingResponse getMeeting(UUID meetingId, Long requesterId) {
        log.info("Retrieving meeting: {} by requester: {}", meetingId, requesterId);
        Meeting meeting = meetingRepository.findById(meetingId)
                .orElseThrow(() -> new AppException(ErrorCode.MEETING_NOT_FOUND));

        // Check membership
        if (!meetingMemberRepository.existsByMeetingIdAndUserId(meetingId, requesterId)) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        return mapToMeetingResponse(meeting);
    }

    public Page<MeetingResponse> listMeetings(Long userId, Pageable pageable) {
        log.info("Listing meetings for user: {}", userId);
        Page<Meeting> meetings = meetingRepository.findMeetingsByUserId(userId, pageable);
        return meetings.map(this::mapToMeetingResponse);
    }

    @Transactional
    public MeetingResponse updateMeeting(UUID meetingId, MeetingUpdateRequest request, Long requesterId) {
        log.info("Updating meeting: {} by requester: {}", meetingId, requesterId);
        Meeting meeting = meetingRepository.findById(meetingId)
                .orElseThrow(() -> new AppException(ErrorCode.MEETING_NOT_FOUND));

        // Requester must be HOST or EDITOR to update meeting
        MeetingMember member = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, requesterId)
                .orElseThrow(() -> new AppException(ErrorCode.UNAUTHORIZED));

        if (member.getRole() != MeetingRole.HOST && member.getRole() != MeetingRole.EDITOR) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        meeting.setTitle(request.getTitle());
        meeting.setDescription(request.getDescription());
        if (request.getStatus() != null) {
            meeting.setStatus(request.getStatus());
        }

        Meeting updatedMeeting = meetingRepository.save(meeting);
        return mapToMeetingResponse(updatedMeeting);
    }

    @Transactional
    public void deleteMeeting(UUID meetingId, Long requesterId) {
        log.info("Deleting meeting: {} by requester: {}", meetingId, requesterId);
        Meeting meeting = meetingRepository.findById(meetingId)
                .orElseThrow(() -> new AppException(ErrorCode.MEETING_NOT_FOUND));

        // Requester must be HOST to delete meeting
        MeetingMember member = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, requesterId)
                .orElseThrow(() -> new AppException(ErrorCode.UNAUTHORIZED));

        if (member.getRole() != MeetingRole.HOST) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        // Delete all members first
        List<MeetingMember> members = meetingMemberRepository.findByMeetingId(meetingId);
        meetingMemberRepository.deleteAll(members);

        meetingRepository.delete(meeting);
    }

    @Transactional
    public MeetingMemberResponse addMember(UUID meetingId, MemberAddRequest request, Long requesterId) {
        log.info("Adding member (userId: {}, email: {}) to meeting: {} by requester: {}", request.getUserId(), request.getEmail(), meetingId, requesterId);

        // Requester must be HOST to add members
        MeetingMember requester = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, requesterId)
                .orElseThrow(() -> new AppException(ErrorCode.UNAUTHORIZED));

        if (requester.getRole() != MeetingRole.HOST) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        Long targetUserId = request.getUserId();
        if (targetUserId == null && request.getEmail() != null) {
            // Lookup user ID by email via UserClient
            try {
                ApiResponse<UserResponse> userResponse = userClient.getUserByEmail(request.getEmail());
                if (userResponse == null || userResponse.getResult() == null) {
                    throw new AppException(ErrorCode.USER_NOT_FOUND);
                }
                targetUserId = userResponse.getResult().getId();
            } catch (AppException ae) {
                throw ae;
            } catch (Exception e) {
                log.error("Failed to lookup user by email: {}", request.getEmail(), e);
                throw new AppException(ErrorCode.USER_NOT_FOUND);
            }
        } else if (targetUserId != null) {
            // Validate user ID existence via UserClient
            try {
                ApiResponse<UserResponse> userResponse = userClient.getUserById(targetUserId);
                if (userResponse == null || userResponse.getResult() == null) {
                    throw new AppException(ErrorCode.USER_NOT_FOUND);
                }
            } catch (AppException ae) {
                throw ae;
            } catch (Exception e) {
                log.error("Failed to validate user ID: {}", targetUserId, e);
                throw new AppException(ErrorCode.USER_NOT_FOUND);
            }
        }

        if (targetUserId == null) {
            throw new AppException(ErrorCode.INVALID_ACTION);
        }

        // Check if member already exists
        if (meetingMemberRepository.existsByMeetingIdAndUserId(meetingId, targetUserId)) {
            throw new AppException(ErrorCode.MEMBER_ALREADY_EXISTS);
        }

        MeetingMember newMember = MeetingMember.builder()
                .meetingId(meetingId)
                .userId(targetUserId)
                .role(request.getRole())
                .build();

        MeetingMember savedMember = meetingMemberRepository.save(newMember);
        return mapToMemberResponse(savedMember);
    }

    @Transactional
    public MeetingMemberResponse updateMemberRole(UUID meetingId, Long targetUserId, MemberUpdateRequest request, Long requesterId) {
        log.info("Updating member: {} role to {} in meeting: {} by requester: {}", targetUserId, request.getRole(), meetingId, requesterId);

        // Requester must be HOST to update roles
        MeetingMember requester = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, requesterId)
                .orElseThrow(() -> new AppException(ErrorCode.UNAUTHORIZED));

        if (requester.getRole() != MeetingRole.HOST) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        MeetingMember targetMember = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, targetUserId)
                .orElseThrow(() -> new AppException(ErrorCode.MEMBER_NOT_FOUND));

        Meeting meeting = meetingRepository.findById(meetingId)
                .orElseThrow(() -> new AppException(ErrorCode.MEETING_NOT_FOUND));

        // Prevent changing the creator's role (the creator must always be HOST)
        if (targetUserId.equals(meeting.getCreatorId()) && request.getRole() != MeetingRole.HOST) {
            throw new AppException(ErrorCode.INVALID_ACTION);
        }

        // Prevent self-role modification if it degrades the host count
        if (targetUserId.equals(requesterId) && request.getRole() != MeetingRole.HOST) {
            long hostCount = meetingMemberRepository.findByMeetingId(meetingId).stream()
                    .filter(m -> m.getRole() == MeetingRole.HOST)
                    .count();
            if (hostCount <= 1) {
                throw new AppException(ErrorCode.INVALID_ACTION);
            }
        }

        targetMember.setRole(request.getRole());
        MeetingMember updatedMember = meetingMemberRepository.save(targetMember);
        return mapToMemberResponse(updatedMember);
    }

    @Transactional
    public void removeMember(UUID meetingId, Long targetUserId, Long requesterId) {
        log.info("Removing member: {} from meeting: {} by requester: {}", targetUserId, meetingId, requesterId);

        Meeting meeting = meetingRepository.findById(meetingId)
                .orElseThrow(() -> new AppException(ErrorCode.MEETING_NOT_FOUND));

        // Prevent removing the creator of the meeting (creator cannot leave or be removed; they must delete the meeting)
        if (targetUserId.equals(meeting.getCreatorId())) {
            throw new AppException(ErrorCode.INVALID_ACTION);
        }

        // Requester must be HOST or they are removing themselves (leaving)
        boolean isSelf = targetUserId.equals(requesterId);
        MeetingMember requester = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, requesterId)
                .orElseThrow(() -> new AppException(ErrorCode.UNAUTHORIZED));

        if (!isSelf && requester.getRole() != MeetingRole.HOST) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        MeetingMember targetMember = meetingMemberRepository.findByMeetingIdAndUserId(meetingId, targetUserId)
                .orElseThrow(() -> new AppException(ErrorCode.MEMBER_NOT_FOUND));

        // If a HOST is leaving, verify they are not the sole HOST
        if (isSelf && targetMember.getRole() == MeetingRole.HOST) {
            long hostCount = meetingMemberRepository.findByMeetingId(meetingId).stream()
                    .filter(m -> m.getRole() == MeetingRole.HOST)
                    .count();
            if (hostCount <= 1) {
                throw new AppException(ErrorCode.INVALID_ACTION);
            }
        }

        meetingMemberRepository.delete(targetMember);
    }

    public List<MeetingMemberResponse> getMembers(UUID meetingId, Long requesterId) {
        log.info("Retrieving members of meeting: {} by requester: {}", meetingId, requesterId);

        // Check if requester is a member of the meeting
        if (!meetingMemberRepository.existsByMeetingIdAndUserId(meetingId, requesterId)) {
            throw new AppException(ErrorCode.UNAUTHORIZED);
        }

        return meetingMemberRepository.findByMeetingId(meetingId).stream()
                .map(this::mapToMemberResponse)
                .collect(Collectors.toList());
    }
}

File: services_ms/apps/meeting/src/meeting.service.ts
import { Injectable } from '@nestjs/common';
import { AppException, ErrorCodes } from '../../../libs/common/src/exceptions/error-code';
import { MeetingRepository } from './repositories/meeting.repository';
import { UserGateway } from './gateways/user.gateway';
import { FileGateway } from './gateways/file.gateway';
import { TranscriptGateway } from './gateways/transcript.gateway';
import { CreateMeetingDto } from './dto/create-meeting.dto';
import { UpdateMeetingDto } from './dto/update-meeting.dto';
import { AddMemberDto } from './dto/add-member.dto';
import { UpdateMemberDto } from './dto/update-member.dto';
import { MeetingRole } from '@prisma/client';

@Injectable()
export class MeetingService {
  constructor(
    private readonly meetingRepo: MeetingRepository,
    private readonly userGateway: UserGateway,
    private readonly fileGateway: FileGateway,
    private readonly transcriptGateway: TranscriptGateway,
  ) {}

  private async enrichMeeting(
    meeting: any,
    includeAudioFile: boolean,
    includeTranscript: boolean,
  ) {
    const response = {
      ...meeting,
      audioFile: null,
      transcript: null,
    };
    if (!meeting.audioFileId) return response;

    if (includeAudioFile) {
      try {
        const fileMetadata = await this.fileGateway.getFileMetadata(
          meeting.audioFileId,
        );
        if (fileMetadata) {
          response.audioFile = fileMetadata;
        }
      } catch (err) {
        console.warn(
          `Failed to fetch audio file metadata for ID: ${meeting.audioFileId}`,
          err.message,
        );
      }
    }

    if (includeTranscript) {
      try {
        const transcript =
          await this.transcriptGateway.getTranscriptByAudioFile(
            meeting.audioFileId,
          );
        if (transcript) {
          response.transcript = transcript;
        }
      } catch (err) {
        console.warn(
          `Failed to fetch transcript for file ID: ${meeting.audioFileId}`,
          err.message,
        );
      }
    }

    return response;
  }

  async createMeeting(dto: CreateMeetingDto, creatorId: number) {
    console.log(
      `Creating meeting with title: ${dto.title}, creatorId: ${creatorId}, audioFileId: ${dto.audioFileId}`,
    );

    // 1. Validate audio file existence using FileGateway
    try {
      const fileExists = await this.fileGateway.checkFileExists(
        dto.audioFileId,
      );
      if (!fileExists) {
        throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
      }
    } catch (err) {
      if (err instanceof AppException) throw err;
      console.error('Failed to connect to file-service for validation', err);
      throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
    }

    // 2. Check if audio file is already linked to another meeting
    const linked = await this.meetingRepo.findByAudioFileId(dto.audioFileId);
    if (linked) {
      throw new AppException(ErrorCodes.AUDIO_FILE_ALREADY_LINKED);
    }

    // 3. Save meeting and add creator as HOST
    return this.meetingRepo.createMeetingWithCreator(
      dto.title,
      dto.description,
      creatorId,
      dto.audioFileId,
    );
  }

  async getMeeting(
    meetingId: string,
    requesterId: number,
    includeAudioFile: boolean,
    includeTranscript: boolean,
  ) {
    console.log(
      `Retrieving meeting: ${meetingId} by requester: ${requesterId}`,
    );
    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // Check membership
    const membership = await this.meetingRepo.findMember(
      meetingId,
      requesterId,
    );
    if (!membership) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    return this.enrichMeeting(meeting, includeAudioFile, includeTranscript);
  }

  async listMeetings(
    userId: number,
    page: number,
    size: number,
    includeAudioFile: boolean,
    includeTranscript: boolean,
  ) {
    console.log(`Listing meetings for user: ${userId}`);
    const skip = page * size;
    const take = size;

    const [meetings, total] = await Promise.all([
      this.meetingRepo.findManyByUserId(userId, skip, take),
      this.meetingRepo.countByUserId(userId),
    ]);

    const enrichedContent = await Promise.all(
      meetings.map((m) =>
        this.enrichMeeting(m, includeAudioFile, includeTranscript),
      ),
    );

    const totalPages = Math.ceil(total / size);

    return {
      content: enrichedContent,
      totalElements: total,
      pageNumber: page,
      pageSize: size,
      totalPages: totalPages,
    };
  }

  async updateMeeting(
    meetingId: string,
    dto: UpdateMeetingDto,
    requesterId: number,
  ) {
    console.log(`Updating meeting: ${meetingId} by requester: ${requesterId}`);
    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // Requester must be HOST or EDITOR to update meeting
    const member = await this.meetingRepo.findMember(meetingId, requesterId);
    if (
      !member ||
      (member.role !== MeetingRole.HOST && member.role !== MeetingRole.EDITOR)
    ) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    const data: any = {
      title: dto.title,
      description: dto.description,
    };
    if (dto.status) {
      data.status = dto.status;
    }
    if (dto.audioFileId) {
      try {
        const fileExists = await this.fileGateway.checkFileExists(
          dto.audioFileId,
        );
        if (!fileExists) {
          throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
        }
      } catch (err) {
        if (err instanceof AppException) throw err;
        console.error('Failed to connect to file-service for validation', err);
        throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
      }

      const linked = await this.meetingRepo.findByAudioFileId(dto.audioFileId);
      if (linked && linked.id !== meetingId) {
        throw new AppException(ErrorCodes.AUDIO_FILE_ALREADY_LINKED);
      }

      data.audioFileId = dto.audioFileId;
    }

    return this.meetingRepo.update(meetingId, data);
  }

  async deleteMeeting(meetingId: string, requesterId: number) {
    console.log(`Deleting meeting: ${meetingId} by requester: ${requesterId}`);
    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // Requester must be HOST to delete meeting
    const member = await this.meetingRepo.findMember(meetingId, requesterId);
    if (!member || member.role !== MeetingRole.HOST) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    await this.meetingRepo.delete(meetingId);

    return { message: 'Meeting deleted successfully' };
  }

  async getMembers(meetingId: string, requesterId: number) {
    console.log(
      `Retrieving members of meeting: ${meetingId} by requester: ${requesterId}`,
    );

    // Check membership of requester
    const requesterMembership = await this.meetingRepo.findMember(
      meetingId,
      requesterId,
    );
    if (!requesterMembership) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    return this.meetingRepo.findMembersByMeetingId(meetingId);
  }

  async addMember(meetingId: string, dto: AddMemberDto, requesterId: number) {
    console.log(
      `Adding member (userId: ${dto.userId}, email: ${dto.email}) to meeting: ${meetingId} by requester: ${requesterId}`,
    );

    // Requester must be HOST to add members
    const requester = await this.meetingRepo.findMember(meetingId, requesterId);
    if (!requester || requester.role !== MeetingRole.HOST) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    let targetUserId = dto.userId;
    if (!targetUserId && dto.email) {
      try {
        const user = await this.userGateway.findProfileByEmail(dto.email);
        if (!user || !user.id) {
          throw new AppException(ErrorCodes.USER_NOT_FOUND);
        }
        targetUserId = user.id;
      } catch (err) {
        throw new AppException(ErrorCodes.USER_NOT_FOUND);
      }
    } else if (targetUserId) {
      try {
        const user = await this.userGateway.findOneProfile(targetUserId);
        if (!user || !user.id) {
          throw new AppException(ErrorCodes.USER_NOT_FOUND);
        }
      } catch (err) {
        throw new AppException(ErrorCodes.USER_NOT_FOUND);
      }
    }

    if (!targetUserId) {
      throw new AppException(ErrorCodes.INVALID_ACTION);
    }

    // Check if member already exists
    const existing = await this.meetingRepo.findMember(meetingId, targetUserId);
    if (existing) {
      throw new AppException(ErrorCodes.MEMBER_ALREADY_EXISTS);
    }

    return this.meetingRepo.addMember(meetingId, targetUserId, dto.role);
  }

  async updateMemberRole(
    meetingId: string,
    targetUserId: number,
    dto: UpdateMemberDto,
    requesterId: number,
  ) {
    console.log(
      `Updating member: ${targetUserId} role to ${dto.role} in meeting: ${meetingId} by requester: ${requesterId}`,
    );

    // Requester must be HOST to update roles
    const requester = await this.meetingRepo.findMember(meetingId, requesterId);
    if (!requester || requester.role !== MeetingRole.HOST) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    const targetMember = await this.meetingRepo.findMember(
      meetingId,
      targetUserId,
    );
    if (!targetMember) {
      throw new AppException(ErrorCodes.MEMBER_NOT_FOUND);
    }

    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // Prevent changing the creator's role (the creator must always be HOST)
    if (targetUserId === meeting.creatorId && dto.role !== MeetingRole.HOST) {
      throw new AppException(ErrorCodes.INVALID_ACTION);
    }

    // Prevent self-role modification if it degrades the host count
    if (targetUserId === requesterId && dto.role !== MeetingRole.HOST) {
      const members = await this.meetingRepo.findMembersByMeetingId(meetingId);
      const hostCount = members.filter(
        (m) => m.role === MeetingRole.HOST,
      ).length;
      if (hostCount <= 1) {
        throw new AppException(ErrorCodes.INVALID_ACTION);
      }
    }

    return this.meetingRepo.updateMemberRole(meetingId, targetUserId, dto.role);
  }

  async removeMember(
    meetingId: string,
    targetUserId: number,
    requesterId: number,
  ) {
    console.log(
      `Removing member: ${targetUserId} from meeting: ${meetingId} by requester: ${requesterId}`,
    );

    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // Prevent removing the creator of the meeting (creator cannot leave or be removed)
    if (targetUserId === meeting.creatorId) {
      throw new AppException(ErrorCodes.INVALID_ACTION);
    }

    const isSelf = targetUserId === requesterId;
    const requester = await this.meetingRepo.findMember(meetingId, requesterId);
    if (!requester) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    if (!isSelf && requester.role !== MeetingRole.HOST) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    const targetMember = await this.meetingRepo.findMember(
      meetingId,
      targetUserId,
    );
    if (!targetMember) {
      throw new AppException(ErrorCodes.MEMBER_NOT_FOUND);
    }

    // If a HOST is leaving, verify they are not the sole HOST
    if (isSelf && targetMember.role === MeetingRole.HOST) {
      const members = await this.meetingRepo.findMembersByMeetingId(meetingId);
      const hostCount = members.filter(
        (m) => m.role === MeetingRole.HOST,
      ).length;
      if (hostCount <= 1) {
        throw new AppException(ErrorCodes.INVALID_ACTION);
      }
    }

    await this.meetingRepo.removeMember(meetingId, targetUserId);

    return { message: 'Member removed from meeting successfully' };
  }

  async getAudioFileId(meetingId: string) {
    const meeting = await this.meetingRepo.findById(meetingId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    if (!meeting.audioFileId) {
      throw new Error('Meeting has no audioFileId');
    }

    return { audioFileId: meeting.audioFileId };
  }

  async getMeetingByAudioFileId(
    audioFileId: string,
    requesterId: number,
    includeAudioFile: boolean,
    includeTranscript: boolean,
  ) {
    console.log(
      `Retrieving meeting by audioFileId: ${audioFileId} by requester: ${requesterId}`,
    );

    // 1. Validate audio file existence using FileGateway
    try {
      const fileExists = await this.fileGateway.checkFileExists(audioFileId);
      if (!fileExists) {
        throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
      }
    } catch (err) {
      if (err instanceof AppException) throw err;
      console.error('Failed to connect to file-service for validation', err);
      throw new AppException(ErrorCodes.AUDIO_FILE_NOT_FOUND);
    }

    // 2. Find meeting associated with this audio file
    const meeting = await this.meetingRepo.findByAudioFileId(audioFileId);
    if (!meeting) {
      throw new AppException(ErrorCodes.MEETING_NOT_FOUND);
    }

    // 3. Verify requester is a member of the meeting
    const membership = await this.meetingRepo.findMember(meeting.id, requesterId);
    if (!membership) {
      throw new AppException(ErrorCodes.UNAUTHORIZED);
    }

    // 4. Enrich meeting information
    return this.enrichMeeting(meeting, includeAudioFile, includeTranscript);
  }
}

